docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,540 @@
+
+*   the variable v is promotable in the statements (S) in the immediately
+    enclosing block that follow the do statement,
+*   e shows that v has the type T when false, and
+*   s is neither a break statement, nor a block containing, directly or
+    indirectly, a break statement,
+
+then v is known to have type T in S.
+
+### 19.1.1 Type Promotion
+
+The static type system ascribes a static type to every expression. In some
+cases, the types of local variables and formal parameters may be promoted from
+their declared types based on control flow.
+
+We say that a variable v is known to have type T whenever we allow the type of v
+to be promoted. The exact circumstances when type promotion is allowed are given
+in the relevant sections of the specification (16.20, 16.22, 16.29, 17.5,
+17.6.1, and 17.7). Each of these sections defines the scope (range of code) in
+which the variable's type is to be promoted.
+
+Type promotion for a variable v is allowed only when we can deduce that such
+promotion is valid based on an analysis of certain boolean expressions. In such
+cases, we either say that the boolean expression b shows that v has type T when
+b is true, or that the boolean expression b shows that v has type T when b is
+false. As a rule, for all variables v and types T, a boolean expression does not
+show that v has type T. Those situations where an expression does show that a
+variable has a type are mentioned explicitly in the relevant sections of this
+specification (16.34 and 16.22).
+
+Dart 1.0: Type promotion for a variable v is also allowed only for a subset of
+variables and is dependent on the use of the variable in the scope in which the
+promotion would occur. If a variable is one for which type promotion is allowed,
+we say that the variable is promotable in some scope. A variable v is promotable
+in the scope S if all of the following conditions are met:
+
+*   v is either a local variable or a parameter,
+
+*   v cannot be potentially mutated within any closure unless all of the types
+    of the values that would be assigned to v within S are more specific than
+    the promoted type of v, and
+
+*   if v is captured by a closure in S then v cannot be potentially mutated
+    anywhere unless all of the types of the values that would be assigned to v
+    within S are more specific than the promoted type of v.
+
+Dart 2.0: Type promotion for a variable v is also allowed only for a subset of
+variables and is dependent on the use of the variable in the scope in which the
+promotion would occur. If a variable is one for which type promotion is allowed,
+we say that the variable is promotable in some scope. A variable v is promotable
+in the scope S if all of the following conditions are met:
+
+*   v is either a local variable or a parameter,
+
+*   v cannot be potentially mutated within S unless all of the types of the
+    values that would be assigned to v within S are a subtype of the promoted
+    type of v,
```

Comment:
I worry (and I think Brian has evidence to support it) that changing (by demoting or refining) the type on assignment leads to a more confusing user experience, even if we sidestep intersection types.

For now, at least, I like Leaf's suggestion of having the assignment simply not change the type. Instead, we type check the assigned value against the variable's (now promoted) type and report an error as usual if that's not valid. So, in your example, you'd get an error on `x = x.length` because you're trying to assign an int to a variable of (promoted) type String.